modules/core/src/main/java/org/apache/ignite/internal/processors/cache/distributed/dht/preloader/GridDhtPreloader.java

```diff
@@ -172,10 +172,12 @@ private IgniteCheckedException stopError() {
 
         RebalanceFuture rebalanceFuture = (RebalanceFuture)rebalanceFuture();
 
-        if (rebalanceFuture.isInitial())
+        if (rebalanceFuture.isInitial() && !rebalanceFuture.isDone())
             return true;
 
-        AffinityTopologyVersion rebTopVer = rebalanceFuture.topologyVersion();
+        AffinityTopologyVersion rebTopVer = rebalanceFuture.isInitial() ?
```

Comment:
I think rebalanceRequired can be simplified to:

/** {@inheritDoc} */
    @Override public boolean rebalanceRequired(GridDhtPartitionsExchangeFuture exchFut) {
        if (ctx.kernalContext().clientNode())
            return false; // No-op.

        AffinityTopologyVersion lastAffChangeTopVer =
            ctx.exchange().lastAffinityChangedTopologyVersion(exchFut.topologyVersion());

        // If this exchange changes affinity rebalance is required.
        return lastAffChangeTopVer.equals(exchFut.topologyVersion());
    }

Reply:
You are use exchFut.topologyVersion() instead of previous version, it cannot be correct solution.
I think this condition will return true in all cases except when exchange which does not change affinity.